sonic_platform_base/chassis_base.py

```diff
@@ -0,0 +1,94 @@
+#
+# chassis_base.py
+#
+# Abstract base class for implementing a platform-specific class with which
+# to interact with a chassis device in SONiC.
+#
+
+try:
+    import abc
+    import sys
+    from . import device_base
+except ImportError as e:
+    raise ImportError(str(e) + " - required module not found")
+
+
+class ChassisBase(device_base.DeviceBase):
+    """
+    Abstract base class for interfacing with a platform chassis
+    """
+
+    __metaclass__ = abc.ABCMeta
+
+    # Possible reboot causes
+    REBOOT_CAUSE_POWER_LOSS = "power_loss" 
+    REBOOT_CAUSE_THERMAL_OVERLOAD = "thermal_overload"
+    REBOOT_CAUSE_INSUFFICIENT_FAN = "insufficient_fan"
+    REBOOT_CAUSE_WATCHDOG = "watchdog"
+    REBOOT_CAUSE_SOFTWARE = "software"
+
+    # List of all fans available on the chassis
+    fan_list = []
+
+    @abc.abstractmethod
+    def get_base_mac(self):
+        """
+        Retrieves the base MAC address for the chassis
+
+        Returns:
+            A string containing the MAC address in the format
+            'XX:XX:XX:XX:XX:XX'
+        """
+        return None 
+
+    @abc.abstractmethod
+    def get_reboot_cause(self):
```

Comment:
> Reboot cause can be persistent on some product and have to be explicitly cleared.
> Otherwise reading this information after a reboot will show all the reload causes that happened previous times.
> 
> I would suggest either a default parameter `clear=False` to this method or to add a new one called `clear_reboot_cause`.
> Also I think this function should only be called once with `clear=True` and the result cached somewhere in a tmpfs. This means that future asks for reload cause information would go through the cache.

Since clearing the reboot cause is platform-specific functionality, I believe this logic should be left to the platform-specific implementation of the function. If the platform requires clearing the reboot cause, it should be done in the implementation and a file saved to tmpfs. The function can check for the presence of the file first before attempting to read/clear the register. I prefer this to adding a `clear=True` parameter to the base method, and I think adding a `clear_reboot_cause()` method is unnecessary, because there should be no need to explicitly clear the reboot cause.

> Some platform might also have limitation on the number of reboot causes that can be stored.

If you're concerned that not all platforms will support all reboot causes, that is expected. The base class will define all possible reboot causes that SONiC understands; the platform-specific implementation will simply return the subset which is applicable to that platform. If this doesn't address your concern, please elaborate.

Reply:
> > Reboot cause can be persistent on some product and have to be explicitly cleared.
> > Otherwise reading this information after a reboot will show all the reload causes that happened previous times.
> > I would suggest either a default parameter `clear=False` to this method or to add a new one called `clear_reboot_cause`.
> > Also I think this function should only be called once with `clear=True` and the result cached somewhere in a tmpfs. This means that future asks for reload cause information would go through the cache.
> 
> Since the need to clear the reboot cause after reading is platform-specific functionality, I believe this logic should be left to the platform-specific implementation of the function. If the platform requires clearing the reboot cause, it should be done in the implementation and a file saved to tmpfs. The function can check for the presence of the file first before attempting to read/clear the register. I prefer this to adding a `clear=True` parameter to the base method, and I think adding a `clear_reboot_cause()` method is unnecessary, because there should be no need to explicitly clear the reboot cause.
> 

I would assume all the platforms to behave the same way with regard to the reboot causes. However this information would have to be crowd-sourced from other vendors as well for more relevant data.
I'm fine with letting the responsibility of clearing and caching the reboot cause to the plugin implementation.
Is there a place in SONiC where such cache files should go? Currently almost everything is backed by the flash including `/tmp`. The only place that is a `tmpfs` and would feel "suitable" seems to be `/run`. Is there some guidelines somewhere regarding this? I would be interested to know the reason why `/tmp` is not a `tmpfs`.

> > Some platform might also have limitation on the number of reboot causes that can be stored.
> 
> If you're concerned that not all platforms will support all reboot causes, that is expected. The base class will define all possible reboot causes that SONiC understands; the platform-specific implementation will simply return the subset which is applicable to that platform. If this doesn't address your concern, please elaborate.

My concern was a bit different. Your previous answer ended up answered my question but I'll elaborate a bit more.
Basically the concept of a reboot cause is to be stored on an external device until being read.
These devices usually have space restriction and can only save a given number of reboot cause after which they won't store anything until being cleared. For example if your device can store 2 reboot cause and you trigger 2 watchdogs and then 1 powerloss, you might not see the powerloss if you haven't cleared the reboot cause after each reboot.